README.md
# CMSC742 Distinguishing Texts Generated by ChatGPT from Human Generated
By Kyle Hassold, Samantha Tang, and Christopher Tharratt.

Please note, for this project you will need to download the Reddit Cross-Topic Authorship Verification Corpus from https://data.mendeley.com/datasets/hppkn5kbg8/1. 

## Additional GPT Author

We discussed in our paper that we added a ChatGPT author to the Reddit Cross-Topic Authorship Verification Corpus. You will find the prompts used to generate the 4 files for this author at `ChatGPT_User_Documents/prompts.txt`. You will find the 4 files in the `ChatGPT_User_Documents` sub-directory as well. 

## Generating GPT Mimics 

To run the python file for generating the GPT Mimics, you will need to install the openai libraries. 
`pip3 install openai` should be all you need to install the appropriate packages, but if you have issues you can look here: https://platform.openai.com/docs/api-reference?lang=python.

To generate the mimics, you will need to change a few variables within the file `mimic.py`. First, set the `openai.api_key` variable to your OpenAI api key. Next, set `path_to_users` on line 7 to the filepath to wherever you have saved the Reddit Cross-Topic Authorship Verification Corpus. Then, set `output_directory` on line 8 to the filepath to where you would like the mimics to be saved. 

Then you can simply run `python3 mimic.py`. 

If any mimics were unsuccessfully created, you can delete the generated mimic in the set `output_directory` rerun the python script as the file checks if a mimic exists alreadt for a given author. You can also change the prompt for that author to one of the other four prompt endings found in the python file. 

You can see the mimic files we generated in the `GPT_Mimics` subdirectory. 

## Fine-tuning Contriever

Note, for the following directions, `Data/RedditData` refers to a directory that was created after the mimics were generated. `Data/RedditData` has a folder for each of the 1001 authors which contains the 4 known documents from the original Reddit Cross-Topic Authorship Verification Corpus (or the 4 generated documents we created in the case of the added GPT Author), 1 unknown document if applicable (applicable for all authors except for GPT Author), and the mimic for each author in the original reddit corpus.

1. Read through the `src/options.py` file to comprehend what values will be used by the fine-tuning script.
2. Run ```python finetuning.py [options]``` with any option values.
   - Example: ```python --model_path=facebook/contriever --train_data=Data/RedditData --eval_data=Data/RedditData --per_gpu_batch_size=16```
   - Note that in the above example, the program is created to be able to parse out the correct number of files to be used for training and for evaluation from the `Data/RedditData` directory. 
3. The training will commence and store a model in the designated location (found in the `src/options.py` file) if options are set correctly.

## Evaluation

1. Generate embeddings for the entire dataset using `gen_embeddings.py` with arguments for the filepaths to the saved model, dataset, and output directory.
   - Example: ```python gen-embeddings.py ./saved_model_directory/saved_model.zip ./RedditData ./output_here```
2. Run evaluation script `eval.py` on the embeddings dataset generated in the previous step
   - There are various evaluation functions and options that can be selected at the end of the file.


ChatGPT_User_Documents/news.txt
It's frustrating to see how much political polarization has infiltrated the vaccination debate. At the end of the day, we all want to protect ourselves and our communities. Let's focus on the science and work together to end this pandemic. This story is a reminder of the ongoing struggle for LGBTQ+ rights around the world. It's important that we continue to fight for equality and support those who are still facing discrimination. I'm glad to see more attention being given to mental health issues, but we still have a long way to go in terms of creating a system that provides accessible and affordable care for everyone who needs it. The rise in anti-Asian hate crimes is deeply concerning. We need to stand in solidarity with our Asian-American friends and neighbors and work to combat xenophobia and racism in all its forms. The fact that we're still grappling with the effects of climate change and environmental degradation in 2023 is unacceptable. We need to take bold action now to protect our planet and ensure a livable future for generations to come. As someone who works in healthcare, I can tell you that the strain on our system during the pandemic has been overwhelming. We need to prioritize the well-being of our healthcare workers and invest in systems that will help us respond more effectively to future crises. It's disheartening to see the continued erosion of press freedom around the world. A free and independent press is crucial to holding those in power accountable and ensuring that the public has access to accurate information. I'm encouraged to see more companies taking steps to address income inequality and improve working conditions for their employees. This is a positive step forward for the labor movement. It's troubling to see the rise of authoritarianism and anti-democratic forces in many parts of the world. We need to stand up for democracy and the rule of law, both at home and abroad. The ongoing refugee crisis is a humanitarian disaster. We need to work together as a global community to provide aid and support to those who have been forced to flee their homes. I'm heartbroken by the recent surge in gun violence across the country. We need to take meaningful action to address this epidemic and ensure that our communities are safe. As a parent, I'm deeply concerned about the impact that social media and online content can have on young people. We need to do more to protect our children from harmful and inappropriate material. The recent protests against police brutality and racial injustice have sparked important conversations about systemic racism in our society. We need to continue to listen to and amplify the voices of those who have been marginalized and oppressed. It's important to remember that mental health issues can affect anyone, regardless of their background or circumstances. We need to create a culture of acceptance and support around mental health and prioritize access to care. The ongoing conflict in Syria is a tragedy that has displaced millions of people and left countless others dead or injured. We need to work toward a peaceful resolution and provide aid and support to those who have been affected by the conflict. The rise of extremist groups and hate speech online is deeply concerning. We need to work together to combat online radicalization and promote civil discourse and respect for different perspectives. The fact that we still have a gender pay gap in 2023 is outrageous. We need to prioritize equal pay for equal work and support policies that promote gender equity in all aspects of society. The continued criminalization of marijuana is a waste of resources and a violation of individual liberties. We need to legalize and regulate marijuana at the federal level and prioritize drug policy reform. It's shocking to see the rise in hate crimes against the Asian-American community. We need to stand in solidarity with them and work towards eradicating racism in all its forms. The issue of police brutality and excessive use of force is not new, but it's disheartening to see it continue to happen. We need to hold law enforcement accountable and push for meaningful reform. It's inspiring to see young people take action on climate change and demand that our leaders take it seriously. We need to support and amplify their voices. The ongoing debate over immigration reform is complex, but we need to remember that we're talking about people's lives and families. We need to find a compassionate and just solution. The issue of income inequality is not just an economic issue, but a moral one. We need to prioritize policies that promote fairness and equity for all. The opioid epidemic has devastated communities across the country. We need to address the root causes of addiction and provide better resources for those struggling with substance abuse. The fact that we still have a gender wage gap in 2023 is unacceptable. We need to prioritize equal pay for equal work and support policies that promote gender equity in all industries. The issue of gun violence is complex, but it's clear that we need to take action to prevent more senseless tragedies from happening. We need to work towards common-sense gun laws that protect our communities. The COVID-19 pandemic has exposed many flaws in our healthcare system, including the need for more accessible and affordable care for all. We need to work towards a more equitable and just healthcare system. It's concerning to see the rise of authoritarian leaders around the world. We need to prioritize democracy and human rights, both at home and abroad. The ongoing conflict in the Middle East is devastating for the people who live there. We need to work towards a peaceful resolution and provide aid and support to those affected by the conflict. The issue of police reform is a complex one, but we need to listen to the voices of those who have been disproportionately affected by police brutality and work towards meaningful change. The fac

eval.py
### Imports ###

import os
import torch
import random
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm
from glob import glob

from src.CustomData import Dataset


### Functions ###

def load_embeddings(src):
    embs = []
    for auth in os.listdir(src):
        for doc in os.listdir(os.path.join(src, auth)):
            if doc.startswith('known'):
                embs.append(torch.load(os.path.join(src, auth, doc)))
    return torch.stack(embs).cpu().detach()


def load_embeddings_split(src, include_gpt=True):
    embs_train, embs_eval, embs_mimics = [], [], []
    for i, auth in enumerate(os.listdir(src)):
        # Ignore ChatGPT if requested
        if not include_gpt and auth == 'ChatGPT':
            continue
        for doc in sorted(os.listdir(os.path.join(src, auth))):
            # Collect embeddings
            emb = torch.load(os.path.join(src, auth, doc))

            # Store in associated list
            if doc.startswith('known'):
                if len(embs_train) < (i+1)*3:
                    embs_train.append(emb)
                else:
                    embs_eval.append(emb)
            elif doc.startswith('gpt'):
                embs_mimics.append(emb)
    # Stack and return embeddings
    return torch.stack(embs_train).cpu().detach(), torch.stack(embs_eval).cpu().detach(),\
        torch.stack(embs_mimics).cpu().detach()


def full_vs_train(src, eval_set=False, mean_train=False):
    # Load train and eval embeddings
    train_embs, eval_embs, _ = load_embeddings_split(src)

    if eval_set:
        if mean_train:
            # Compare eval set to the training set with author samples averaged
            scores = eval_embs @ train_embs.reshape(-1, 3, 768).mean(dim=1).T
        else:
            # Compare eval set to the training set
            scores = eval_embs @ train_embs.T
    else:
        # Compare train set with train set removing self comparisons
        scores = train_embs @ train_embs.T
        scores = scores.fill_diagonal_(0)

    # Find max author indexes and associated correct labels
    maxes = torch.argmax(scores, dim=-1) // (3 if not eval_set or (eval_set and not mean_train) else 1)
    labels = torch.arange(0, maxes.size()[0], dtype=torch.long) // (1 if eval_set else 3)

    # Return accuracy
    return (maxes == labels).float().mean().item()


def full_vs_train_recall(src, r, mean_train=False):
    # Collect embeddings
    train_embs, eval_embs, _ = load_embeddings_split(src)

    # Calculate scores
    if mean_train:
        scores = eval_embs @ train_embs.reshape(-1, 3, 768).mean(dim=1).T
    else:
        scores = eval_embs @ train_embs.T

    # Sort based on score and find index of correct labels
    order = torch.argsort(scores, dim=-1, descending=True) // (1 if mean_train else 3)
    labels = torch.arange(0, scores.size()[0], dtype=torch.long)
    indexes = [(o == l).nonzero(as_tuple=True)[0][0] for o, l in zip(order, labels)]

    # Find accuracy for a given recall amount
    if isinstance(r, int):
        return sum(np.array(indexes) < r) / scores.size()[0]
    return [sum(np.array(indexes) < r_i) / scores.size()[0] for r_i in r]


def eval_vs_train_recall_plot(src, save=None):
    # Collect embeddings
    train_embs, eval_embs, _ = load_embeddings_split(src)

    # Calculate scores
    scores = eval_embs @ train_embs.T

    # Sort based on score and find index of correct labels
    order = torch.argsort(scores, dim=-1, descending=True) // 3
    labels = torch.arange(0, scores.size()[0], dtype=torch.long)
    indexes = [(o == l).nonzero(as_tuple=True)[0][0] for o, l in zip(order, labels)]

    # Randomize order and calculate indexes of randomization
    order_rand = torch.clone(order)
    for i in range(order_rand.size(0)):
        order_rand[i] = order_rand[i, torch.randperm(order_rand.size(1))]
    indexes_rand = [(o == l).nonzero(as_tuple=True)[0][0] for o, l in zip(order_rand, labels)]

    # Plot expected cumulative probabilities
    _, _, patches = plt.hist(indexes_rand+[3000], scores.size()[1], density=True, histtype='step',
                             cumulative=True, label='Random')
    patches[0].set_xy(patches[0].get_xy()[:-1])

    # Plot evaluated cumulative probabilities
    _, _, patches = plt.hist(indexes+[3000], scores.size()[1], density=True, histtype='step',
                             cumulative=True, label='Contriever')
    patches[0].set_xy(patches[0].get_xy()[:-1])

    # Make plots pretty
    plt.xlabel('Number of Predictions')
    plt.ylabel('Probability')
    plt.title('One vs One:\nProbability of Correct Author Within N Predictions')
    plt.legend()
    if save:
        plt.savefig(save)
        plt.close()
    else:
        plt.show()


def eval_vs_mean_train_recall_plot(src, save=None):
    # Collect embeddings
    train_embs, eval_embs, _ = load_embeddings_split(src)

    # Calculate scores
    scores = eval_embs @ train_embs.reshape(-1, 3, 768).mean(dim=1).T

    # Sort based on score and find index of correct labels
    order = torch.argsort(scores, dim=-1, descending=True)
    labels = torch.arange(0, scores.size()[0], dtype=torch.long)
    indexes = [(o == l).nonzero(as_tuple=True)[0][0] for o, l in zip(order, labels)]

    # Randomize order and calculate indexes of randomization
    order_rand = torch.clone(order)
    for i in range(order_rand.size(0)):
        order_rand[i] = order_rand[i, torch.randperm(order_rand.size(1))]
    indexes_rand = [(o == l).nonzero(as_tuple=True)[0][0] for o, l in zip(order_rand, labels)]

    # Plot expected cumulative probabilities
    _, _, patches = plt.hist(indexes_rand + [1000], scores.size()[1], density=True, histtype='step',
                             cumulative=True, label='Random')
    patches[0].set_xy(patches[0].get_xy()[:-1])

    # Plot expected and evaluation cumulative probabilities
    n, bins, patches = plt.hist(indexes + [1000], scores.size()[1], density=True, histtype='step',
                                cumulative=True, label='Contriever')
    patches[0].set_xy(patches[0].get_xy()[:-1])

    # Make plots pretty
    plt.xlabel('Number of Predictions')
    plt.ylabel('Probability')
    plt.title('Author vs Author:\nProbability of Correct Author Within N Predictions')
    plt.legend()
    if save:
        plt.savefig(save)
        plt.close()
    else:
        plt.show()


def one_vs_one(src, num=1000):
    # Collect dataset of authors embeddings
    data = Dataset(src)
    count = 0
    for _ in tqdm(range(num)):
        # Choose author #1
        auth = random.randint(0, len(data)-1)
        # Collect random train sample from author #1 for query/anchor
        query = data.get_known(auth)[0]
        query = torch.load(os.path.join(data.data[auth], query))
        # Collect eval sample from author #1 for positive pair
        gold = data.get_known(auth, mask=(3,))[0]
        gold = torch.load(os.path.join(data.data[auth], gold))

        # Collect eval sample from random author #2 for negative pair
        new_auth = random.choice(list(set(range(len(data))) - {auth}))
        neg = data.get_known(new_auth, mask=(3,))[0]
        neg = torch.load(os.path.join(data.data[new_auth], neg))

        # Check if embeddings properly match samples
        if query @ gold > query @ neg:
            count += 1

    # Return accuracy
    return count / num


def auth_vs_auth(src, num=1000, method='max'):
    # Collect dataset of authors embeddings
    data = Dataset(src)
    count = 0
    for _ in tqdm(range(num)):
        # Choose author #1
        auth = random.randint(0, len(data)-1)
        # Collect eval sample from author #1 for query/anchor
        query = data.get_known(auth, mask=(3,))[0]
        query = torch.load(os.path.join(data.data[auth], query))
        # Collect all train samples from author #1 for positive pairs
        golds = data.get_known(auth)
        golds = torch.stack([torch.load(os.path.join(data.data[auth], gold)) for gold in golds])

        # Collect all train samples from random author #2 for negative pairs
        new_auth = random.choice(list(set(range(len(data))) - {auth}))
        negs = data.get_known(new_auth)
        negs = torch.stack([torch.load(os.path.join(data.data[new_auth], neg)) for neg in negs])

        # Evaluate embeddings using certain method
        if method == 'max' and max(query @ golds.T) > max(query @ negs.T):
            # Which author has the highest scoring pair with anchor
            count += 1
        elif method == 'mean' and torch.mean(query @ golds.T) > torch.mean(query @ negs.T):
            # Which author's mean score is highest
            count += 1
        elif method == 'min' and min(query @ golds.T) > min(query @ negs.T):
            # Which author is the highest score based on their worst score
            count += 1

    # Return accuracy
    return count / num


def one_vs_gpt(src, num=1000):
    # Collect training samples from ChatGPT as positive pairs
    golds = [torch.load(os.path.join(src, 'ChatGPT', doc))
             for doc in os.listdir(os.path.join(src, 'ChatGPT'))[:3]]

    # Collect authors without ChatGPT
    auths = os.listdir(src)
    auths.remove('ChatGPT')
    count = 0
    for auth in random.choices(auths, k=num):
        # Collect ChatGPT mimic embedding for the chosen author
        query = torch.load(glob(os.path.join(src, auth, 'gpt_*.txt'))[0])
        # Collect a training sample from chosen author as the negative sample
        docs = sorted(glob(os.path.join(src, auth, 'known*.txt')))[:3]
        random.shuffle(docs)
        neg = torch.load(docs[0])

        # Evaluate if embeddings are accuracy
        if query @ golds[random.randint(0, len(golds)-1)] > query @ neg:
            count += 1
    # Return accuracy
    return count / num


def auth_vs_gpt(src, num=1000, method='max'):
    # Collect training samples from ChatGPT as positive pairs
    golds = torch.stack([torch.load(os.path.join(src, 'ChatGPT', doc))
                         for doc in os.listdir(os.path.join(src, 'ChatGPT'))[:3]])

    # Collect authors without ChatGPT
    auths = os.listdir(src)
    auths.remove('ChatGPT')
    count = 0
    for auth in random.choices(auths, k=num):
        # Collect ChatGPT mimic embedding for the chosen author
        query = torch.load(glob(os.path.join(src, auth, 'gpt_*.txt'))[0])
        # Collect training samples from chosen author as negative samples
        docs = sorted(glob(os.path.join(src, auth, 'known*.txt')))[:3]
        negs = torch.stack([torch.load(doc) for doc in docs])

        # Evaluate if embeddings are accuracy using the chosen method
        if method == 'max' and max(query @ golds.T) > max(query @ negs.T):
            # Which author has the highest scoring pair with anchor
            count += 1
        elif method == 'mean' and torch.mean(query @ golds.T) > torch.mean(query @ negs.T):
            # Which author's mean score is highest
            count += 1
        elif method == 'min' and min(query @ golds.T) > min(query @ negs.T):
            # Which author is the highest score based on their worst score
            count += 1

    # Return accuracy
    return count / num


### Run Code ###

if __name__ == '__main__':
    embedding_src = './Data/RedditEmbeddings2'
    print(f"One vs One accuracy: {one_vs_one(embedding_src, num=10000):.2%}")
    print(f"Author vs Author accuracy: {auth_vs_auth(embedding_src, num=10000):.2%}")
    # print(f"Author Mean vs Author Mean accuracy: {auth_vs_auth(embedding_src, method='mean'):.2%}")
    # print(f"Author Furthest vs Author Furthest accuracy: {auth_vs_auth(embedding_src, method='min'):.2%}")
    print(f"One vs ChatGPT accuracy: {one_vs_gpt(embedding_src, num=10000):.2%}")
    print(f"Author vs ChatGPT accuracy: {auth_vs_gpt(embedding_src, num=10000):.2%}")
    print(f"Author Mean vs ChatGPT Mean accuracy: {auth_vs_gpt(embedding_src, method='mean'):.2%}")
    print(f"Author Furthest vs ChatGPT Furthest accuracy: {auth_vs_gpt(embedding_src, method='min'):.2%}")
    print(f"Train set vs Train set accuracy: {full_vs_train(embedding_src):.2%}")
    print(f"Eval set vs Train set accuracy: {full_vs_train(embedding_src, eval_set=True):.2%}")
    print(f"Eval set vs Mean Train set accuracy: {full_vs_train(embedding_src, eval_set=True, mean_train=True):.2%}")
    r5, r20, r100 = full_vs_train_recall(embedding_src, [5, 20, 100])
    print(f"Eval set vs Train set recall @ 5: {r5:.2%}, 20: {r20:.2%}, 100: {r100:.2%}")
    r5, r20, r100 = full_vs_train_recall(embedding_src, [5, 20, 100], mean_train=True)
    print(f"Eval set vs Mean Train set recall @ 5: {r5:.2%}, 20: {r20:.2%}, 100: {r100:.2%}")
    eval_vs_train_recall_plot(embedding_src, save='./eval_vs_train.png')
    eval_vs_mean_train_recall_plot(embedding_src, save='./eval_vs_mean_train.png')


mimic.py
import openai
import os
import time

openai.api_key =  "" #Set your OpenAI api key here 

path_to_users = "/Users/samanthatang/Documents/UMD/Grad_School/Spring_2023/742/Reddit_Cross-Topic-AV-Corpus_(1000_users)" # Set this to your path to the Reddit Dataset
output_directory = "/Users/samanthatang/Documents/UMD/Grad_School/Spring_2023/742/gpt_texts" # Set this to the path to the folder you would like mimics sent to
# Switch file path to location of Reddit Dataset, get a list of the items in the Dataset directory, then sort the list alphabetically
os.chdir(path_to_users)
dir = os.listdir()
dir_sorted = sorted(dir, key=str.lower)
# Loop through each item in the dataset (At this point, these should be folders for each author)
for i in range(0,len(dir_sorted)):
    # Check you are at a folder as there are other items in the dataset directory
    folder = dir_sorted[i]  
    if os.path.isdir(f"{path_to_users}/{folder}"):
        # Sleep to not overload the model
        time.sleep(10)
        # Move into the specified author's folder, get a list of the items in the folder, then sort the list alphabetically
        os.chdir(f"{path_to_users}/{folder}")
        folder_dir = os.listdir()
        fdir_sorted = sorted(folder_dir, key=str.lower)
        # Pull the 4th item in the folder as it is the 4th known document we will use for evaluation
        file = fdir_sorted[3]
        topic = "[" + file.split("[")[1]
        gpt_file = "gpt_" + folder + topic + ".txt"
        # Form a name for the mimic file
        mimic_check = f"{output_directory}/{gpt_file}"
        if (not os.path.isfile(mimic_check)):
            file_path = f"{path_to_users}/{folder}/{file}"
            # Form the prompt
            with open(file_path, "r") as file:
                prompt = "The following was written by Charlie. \""  
                # Copy in the first 8000 bytes of the text written by the original author
                prompt = prompt + file.read(8000)
                prompt = prompt + "\" Please mimic Charlie's writing style and write on the same topic. You must write at least 800 words."
                # The following are the variations for prompts that could be applied, comment out the above line and uncomment the one you wish to use
                #prompt = prompt + "\" Do not respond to Charlie. However, please mimic Charlie's writing style and write on the same topic. You must write at least 800 words."
                #prompt = prompt + "\" Can you write ramblings in the style of Charlie on the same topics. You must write at least 800 words. "
                #prompt = prompt + "\" DO NOT respond to Charlie. Instead, write in the style of Charlie on the same topics using at least 800 words."
                #prompt = prompt + "\" Please mimic Charlie's writing style and write on the same topic. DO NOT RESPOND TO CHARLIE. JUST ACT AS IF YOU ARE HIM. You must write at least 800 words."
            # Send the formed prompt to GPT
            response = openai.ChatCompletion.create(
                    model="gpt-3.5-turbo",
                    messages = [{"role": "user", "content": prompt}],
                    max_tokens = 2000,
                    temperature = 0)
            print("Finished:", folder)
            # Clean up the response by removing new lines and extraneous white space
            text = response.choices[0].message.content
            text = os.linesep.join([s for s in text.splitlines() if s]).replace("\n", " ")
            # Move into the directory to save mimics in and save the generated response there
            os.chdir(output_directory)
            with open(gpt_file, 'a') as f:
                f.write(text)

gen_embeddings.py
### Imports ###

from tqdm import tqdm
import torch
import sys
import os

from src.contriever import load_retriever


### Functions ###

def gen_embedding(src, encoder, tokenizer):
    # Collect documents in the src folder
    docs = os.listdir(src)
    passages = []
    for doc in docs:
        # Open and save each passage
        with open(os.path.join(src, doc), encoding="utf8") as fin:
            passages.append(fin.read())
    # Convert passages into tokens on the GPU
    tokens = tokenizer(passages, padding=True, truncation=True, return_tensors="pt")
    tokens = {k: v.cuda() for k, v in tokens.items()}
    # Return the list of documents and associated embeddings
    return docs, encoder(**tokens, normalize=True)


def data_loop(src, dist, encoder, tokenizer):
    # Create output folder if needed
    if not os.path.exists(dist):
        os.makedirs(dist)

    # Search through all author folders
    for auth in tqdm(os.listdir(src), file=sys.stdout):
        # Ignore non-directories and folders that have already been processed
        if os.path.isdir(os.path.join(src, auth)) and not os.path.exists(os.path.join(dist, auth)):
            os.makedirs(os.path.join(dist, auth))
            # Generate embeddings and save
            docs, embeddings = gen_embedding(os.path.join(src, auth), encoder, tokenizer)
            for d, e in zip(docs, embeddings):
                torch.save(e, os.path.join(dist, auth, d))


### Run Code ###

if __name__ == '__main__':
    # Path to saved contriever model
    model_path = './checkpoint/exp1/checkpoint/step-5000'
    if len(sys.argv) >= 2:
        model_path = sys.argv[1]

    # Load model and tokenize and move model to GPU
    model, tokenizer, _ = load_retriever(model_path)
    model = model.cuda()
    model.eval()

    # Generate embeddings for dataset
    src = './Data/RedditData'
    dest = './Data/RedditEmbeddings2'
    if len(sys.argv) >= 3:
        src = sys.argv[2]
    if len(sys.argv) >= 4:
        dest = sys.argv[3]

    data_loop(src, dest, model, tokenizer)



src/options.py
# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved

import argparse
import os


class Options:
    def __init__(self):
        self.parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
        self.initialize()

    def initialize(self):
        # basic parameters
        self.parser.add_argument(
            "--output_dir", type=str, default="./checkpoint/my_experiments", help="models are saved here"
        )
        self.parser.add_argument(
            "--train_data",
            nargs="+",
            default=['Data/RedditData'],
            help="Data used for training, passed as a list of directories splitted into tensor files.",
        )
        self.parser.add_argument(
            "--eval_data",
            nargs="+",
            default=['Data/RedditData'],
            help="Data used for evaluation during finetuning, this option is not used during contrastive pre-training.",
        )
        self.parser.add_argument(
            "--eval_datasets", nargs="+", default=[], help="List of datasets used for evaluation, in BEIR format"
        )
        self.parser.add_argument(
            "--eval_datasets_dir", type=str, default="./", help="Directory where eval datasets are stored"
        )
        self.parser.add_argument("--model_path", type=str, default="facebook/contriever", help="path for retraining")
        self.parser.add_argument("--continue_training", action="store_true")
        self.parser.add_argument("--num_workers", type=int, default=1)

        self.parser.add_argument("--chunk_length", type=int, default=256)
        self.parser.add_argument("--loading_mode", type=str, default="split")
        self.parser.add_argument("--lower_case", action="store_true", help="perform evaluation after lowercasing")
        self.parser.add_argument(
            "--sampling_coefficient",
            type=float,
            default=0.0,
            help="coefficient used for sampling between different datasets during training, \
                by default sampling is uniform over datasets",
        )
        self.parser.add_argument("--augmentation", type=str, default="none")
        self.parser.add_argument("--prob_augmentation", type=float, default=0.0)

        self.parser.add_argument("--dropout", type=float, default=0.1)
        self.parser.add_argument("--rho", type=float, default=0.05)

        self.parser.add_argument("--contrastive_mode", type=str, default="moco")
        self.parser.add_argument("--queue_size", type=int, default=65536)
        self.parser.add_argument("--temperature", type=float, default=1.0)
        self.parser.add_argument("--momentum", type=float, default=0.999)
        self.parser.add_argument("--moco_train_mode_encoder_k", action="store_true")
        self.parser.add_argument("--eval_normalize_text", action="store_true")
        self.parser.add_argument("--norm_query", action="store_true")
        self.parser.add_argument("--norm_doc", action="store_true")
        self.parser.add_argument("--projection_size", type=int, default=768)

        self.parser.add_argument("--ratio_min", type=float, default=0.1)
        self.parser.add_argument("--ratio_max", type=float, default=0.5)
        self.parser.add_argument("--score_function", type=str, default="dot")
        self.parser.add_argument("--retriever_model_id", type=str, default="bert-base-uncased")
        self.parser.add_argument("--pooling", type=str, default="average")
        self.parser.add_argument("--random_init", action="store_true", help="init model with random weights")

        # dataset parameters
        self.parser.add_argument("--per_gpu_batch_size", default=20, type=int, help="Batch size per GPU for training.")
        self.parser.add_argument(
            "--per_gpu_eval_batch_size", default=64, type=int, help="Batch size per GPU for evaluation."
        )
        self.parser.add_argument("--total_steps", type=int, default=5000)
        self.parser.add_argument("--warmup_steps", type=int, default=-1)

        self.parser.add_argument("--local_rank", type=int, default=-1, help="For distributed training: local_rank")
        self.parser.add_argument("--main_port", type=int, default=10001, help="Master port (for multi-node SLURM jobs)")
        self.parser.add_argument("--seed", type=int, default=0, help="random seed for initialization")
        # training parameters
        self.parser.add_argument("--optim", type=str, default="adamw")
        self.parser.add_argument("--scheduler", type=str, default="linear")
        self.parser.add_argument("--lr", type=float, default=1e-4, help="learning rate")
        self.parser.add_argument(
            "--lr_min_ratio",
            type=float,
            default=0.0,
            help="minimum learning rate at the end of the optimization schedule as a ratio of the learning rate",
        )
        self.parser.add_argument("--weight_decay", type=float, default=0.01, help="learning rate")
        self.parser.add_argument("--beta1", type=float, default=0.9, help="beta1")
        self.parser.add_argument("--beta2", type=float, default=0.98, help="beta2")
        self.parser.add_argument("--eps", type=float, default=1e-6, help="eps")
        self.parser.add_argument(
            "--log_freq", type=int, default=100, help="log train stats every <log_freq> steps during training"
        )
        self.parser.add_argument(
            "--eval_freq", type=int, default=500, help="evaluate model every <eval_freq> steps during training"
        )
        self.parser.add_argument("--save_freq", type=int, default=1000)
        self.parser.add_argument("--maxload", type=int, default=None)
        self.parser.add_argument("--label_smoothing", type=float, default=0.0)

        # finetuning options
        self.parser.add_argument("--negative_ctxs", type=int, default=1)
        self.parser.add_argument("--negative_hard_min_idx", type=int, default=0)
        self.parser.add_argument("--negative_hard_ratio", type=float, default=0.0)

    def print_options(self, opt):
        message = ""
        for k, v in sorted(vars(opt).items()):
            comment = ""
            default = self.parser.get_default(k)
            if v != default:
                comment = f"\t[default: %s]" % str(default)
            message += f"{str(k):>40}: {str(v):<40}{comment}\n"
        print(message, flush=True)
        model_dir = os.path.join(opt.output_dir, "models")
        if not os.path.exists(model_dir):
            os.makedirs(os.path.join(opt.output_dir, "models"))
        file_name = os.path.join(opt.output_dir, "opt.txt")
        with open(file_name, "wt") as opt_file:
            opt_file.write(message)
            opt_file.write("\n")

    def parse(self):
        opt, _ = self.parser.parse_known_args()
        # opt = self.parser.parse_args()
        return opt
